AIP-86 - Implement basic functionality for deadline alerts async callbacks

ramitkataria opened the pull request:
As discussed in dev calls and [email list](https://lists.apache.org/thread/cpnngjd9vl3wszjczwqkvl13p5pvbm5d), async callbacks for deadline alerts will run in the triggerer. This implements foundational changes needed for the callbacks to run.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

o-nikolas commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +107,33 @@ def _determine_resource() -> tuple[str, str]:
 
+    def handle_miss(self, session: Session):
+        """Handle a missed deadline by creating and starting a trigger to run the callback."""
+        callback_func = import_string(self.callback)
```
This means the callback needs to be both present on the scheduler (where I'm assuming this will be called eventually) _and_ in the Triggerer. Is that a reasonable restriction to put on the users?

ramitkataria replied:
I didn't think about that edge case. Do you have any suggestions for how else I could check to see if a callback is async?

ramitkataria replied:
For now, I've added a TODO to improve this so that this PR doesn't get blocked because of this. What do you think?

o-nikolas replied:
I think that's reasonable, we can chat more about it and iterate for the next PR (or a separate one).

ferruzzi replied:
Yeah, interesting.   The import_string() method was exactly the "decoding" I had in mind when I serialized the callback the way I did, but at the time we weren't thinking about sync vs async.... 

I don't have a ton of experience with async python, but a quick google seems to imply that we could make an async wrapper... maybe a cleaner solution could be to send it all to the Triggerrer and add some logic to the effect of "if it's async then run it, else wrap it before running it" something like

```
async def async_wrapper(callback):
    return await asyncio.to_thread(sync_function, callback)
```

o-nikolas commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -98,6 +107,33 @@ def _determine_resource() -> tuple[str, str]:
+
+        trigger_orm = Trigger.from_object(callback_trigger)
+        session.add(trigger_orm)
+        session.flush()
```
Should the flush come after the second add()?

ramitkataria replied:
That's what I tried earlier but I think this flush is needed before the second add because the trigger's id wouldn't exist before the flush. Also, there's another flush/commit called somewhere along the call stack after the second add so it doesn't need the explicit flush here

o-nikolas commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -64,6 +70,9 @@ class Deadline(Base):
 
     dagrun = relationship("DagRun", back_populates="deadlines")
 
+    trigger_id = Column(Integer, ForeignKey("trigger.id"), nullable=True)
```
I liked the pattern above of including a comment that explains what this is and/or how it's used.

ramitkataria replied:
I agree, fixed!

o-nikolas reviewed (commented):
Few comments so far. As you called out, needs fixes still before final review.

ramitkataria committed:
AIP-86 - Implement basic functionality for deadline alerts async callbacks

As discussed in dev calls and [email list](https://lists.apache.org/thread/cpnngjd9vl3wszjczwqkvl13p5pvbm5d),
async callbacks for deadline alerts will run in the triggerer. This
implements foundational changes needed for the callbacks to run.

ferruzzi commented on airflow-core/src/airflow/models/trigger.py:
```diff
@@ -212,10 +223,10 @@ def clean_unused(cls, session: Session = NEW_SESSION) -> None:
                 )
 
-        # Get all triggers that have no task instances and assets depending on them and delete them
+        # Get all triggers that have no task instances, assets or deadlines depending on them and delete them
```
```suggestion
        # Get all triggers that have no task instances, assets, or deadlines depending on them and delete them
```

nit, sorry.

o-nikolas replied:
I kinda like the extra comma if you want a counter nit, but I know those are both competing styles.

ferruzzi replied:
I was suggesting the extra/oxford comma :P

o-nikolas replied:
Ahhh, okay, perfect. Punctuation nerds unite :smile:

ferruzzi commented on airflow-core/src/airflow/models/trigger.py:
```diff
@@ -366,9 +386,9 @@ def get_sorted_triggers(cls, capacity: int, alive_triggerer_ids: list[int] | Sel
-        return ti_triggers + asset_triggers
+        # Add triggers associated to deadlines first, then tasks, then assets
+        # It prioritizes deadline triggers, then DAGs over event driven scheduling which is fair
+        return deadline_triggers + ti_triggers + asset_triggers
```
Do we need to limit this list to `capacity` length?  We're limiting the deadlines results, but could then go over capacity when adding the others, right?

ramitkataria replied:
I thought about that but I noticed that asset triggers were not accounting for this either so I assumed this is probably the correct way. So I'm not sure but looking into it

ferruzzi replied:
If this is how it is done in other places, I'm willing to be wrong. Feel free to resolve this one but maybe leave it open till the end in case someone else has a thought?

vincbeck replied:
Probably a miss on my side but I think indeed, we should limit this to `capacity`

ramitkataria replied:
Sounds good, I've pushed the change

ferruzzi commented on airflow-core/src/airflow/jobs/triggerer_job_runner.py:
```diff
@@ -600,11 +601,15 @@ def update_triggers(self, requested_trigger_ids: set[int]):
             new_trigger_orm = new_triggers[new_id]
 
-            # If the trigger is not associated to a task or an asset, this means the TaskInstance
+            # If the trigger is not associated to a task, an asset or a deadline, this means the TaskInstance
```
```suggestion
            # If the trigger is not associated to a task, an asset, or a deadline, this means the TaskInstance
```

Sorry, can't help myself...

o-nikolas reviewed (approved):
Src lgtm after comments are resolved. Should we have unit tests for any of this?

ferruzzi reviewed (approved):
My concerns were addressed.  Approved pending merge conflict resolution.

ashb commented on airflow-core/src/airflow/models/trigger.py:
```diff
@@ -192,6 +194,15 @@ def fetch_trigger_ids_with_asset(cls, session: Session = NEW_SESSION) -> set[str
+        from airflow.models.deadline import Deadline
+
+        query = select(Deadline.trigger_id).where(Deadline.trigger_id.is_not(None))
+        return {trigger_id for trigger_id in session.scalars(query)}
```
Hmmmm, I'm not a fan of this pattern -- we now have it in two places, we might need to think about a more general purpose way than adding another special cased column/relationship

ramitkataria replied:
Makes sense, I've made it more general purpose now so anytime there's a new special case similar to this, only a function in the trigger model class would need to be modified

ashb commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -150,6 +160,36 @@ def prune_deadlines(cls, *, session: Session, conditions: dict[Column, Any]) ->
+        # TODO: Improve this check so that the callback doesn't need to be present on the scheduler
+        if not asyncio.iscoroutinefunction(callback_func):
+            # TODO: For sync callbacks, use executor instead of trigger
+            logger.error("Sync callbacks not supported yet: %s", self.callback)
```
Could this be detected at parse time Instead of only when the callback is being run?

ramitkataria replied:
I want to provide the ability for the user to only have the callback present on the host where it will be run (ie the triggerer or executor). In that use case, it would not be possible to detect at parse time. Because of this (among a couple of other reasons), we decided to change the DeadlineAlert SDK definition so that the user would need to specify what kind of callback it is

ashb commented on airflow-core/src/airflow/triggers/deadline.py:
```diff
@@ -0,0 +1,47 @@
+        )
+
+    async def run(self) -> AsyncIterator[TriggerEvent]:
+        callback = import_string(self.callback_path)
```
For this to run, there's a reasonable chance that well need to load the DAG bundle to get the callback, right? I.e. callbacks are defined inside the DAG file?

ramitkataria replied:
The callback function needs to be present on the host somehow and we have a few mechanisms in mind. For the purpose of this PR, the user needs to specify a path that can be imported in the triggerer

ashb replied:
Makes sense for this one -- what is the long term idea here?

ramitkataria replied:
Long term, we want to provide both options - defining it in the DAG file or providing any path importable in the relevant host

ashb reviewed (commented):
Given this a first surface-level pass

ashb commented on airflow-core/src/airflow/jobs/triggerer_job_runner.py:
```diff
@@ -589,7 +589,7 @@ def update_triggers(self, requested_trigger_ids: set[int]):
         # Bulk-fetch new trigger records
         new_triggers = Trigger.bulk_fetch(new_trigger_ids)
-        triggers_with_assets = Trigger.fetch_trigger_ids_with_asset()
+        trigger_ids_with_non_task_associations = Trigger.fetch_trigger_ids_with_non_task_associations()
```
How about?
```suggestion
        non_task_triggers = Trigger.fetch_non_task_trigger_ids()
```

(etc)

ramitkataria replied:
I thought about that earlier as well but I changed it because it may imply that `non_task_triggers` includes all triggers where `task_instance` is `None` (regardless of whether they have an asset or a deadline). Another one I considered was `non_task_active_triggers` but I wasn't sure if calling them "active" is a good idea either because this is before a `workload` is created for them

ramitkataria replied:
I'm interested in getting this merged asap so that the next PRs for deadlines can be created. If you still strongly believe that we should call it `non_task_triggers`, I'm happy to make that change in a later PR but I'll resolve for now
